bindings/go/tests/endpoint_test.go

```diff
@@ -9,81 +9,151 @@ import (
-	defer receiverWorker.Close()
-	workerAddress, _ := receiverWorker.GetAddress()
-	defer workerAddress.Close()
-
-	senderWorker, _ := ucpContext2.NewWorker(ucpWorkerParams)
-	defer senderWorker.Close()
-
-	epParams := (&UcpEpParams{}).SetUcpAddress(workerAddress)
-	epParams.SetPeerErrorHandling().SetErrorHandler(func(ep *UcpEp, status UcsStatus) {
-		t.Logf("Error handler called with status %d", status)
-	})
-	senderEp, _ := senderWorker.NewEndpoint(epParams)
-
-	flushRequest, _ := senderEp.FlushNonBlocking(nil)
-
-	for flushRequest.GetStatus() == UCS_INPROGRESS {
-		senderWorker.Progress()
-		receiverWorker.Progress()
-	}
 
-	flushRequest.Close()
+	for _, memType := range get_mem_types() {
+		sender, receiver := prepare(t)
+		t.Logf("Testing %v -> %v", memType.senderMemType, memType.recvMemType)
 
-	sendMem := CBytes(sendData)
-	defer FreeNativeMemory(sendMem)
-	receiveMem := AllocateNativeMemory(4096)
-	defer FreeNativeMemory(receiveMem)
+		ucpWorkerParams := (&UcpWorkerParams{}).SetThreadMode(UCS_THREAD_MODE_SINGLE)
+		ucpWorkerParams.WakeupTagSend().WakeupTagRecv()
 
-	recvRequest, _ := receiverWorker.RecvTagNonBlocking(receiveMem, 4096, 1, 1, &UcpRequestParams{
-		Cb: func(request *UcpRequest, status UcsStatus, tagInfo *UcpTagRecvInfo) {
+		receiver.worker, _ = receiver.context.NewWorker(ucpWorkerParams)
+		workerAddress, _ := receiver.worker.GetAddress()
 
-			if status != UCS_OK {
-				t.Fatalf("Request failed with status: %d", status)
-			}
+		sender.worker, _ = sender.context.NewWorker(ucpWorkerParams)
+		epParams := (&UcpEpParams{}).SetUcpAddress(workerAddress)
+		epParams.SetPeerErrorHandling().SetErrorHandler(func(ep *UcpEp, status UcsStatus) {
+			t.Logf("Error handler called with status %d", status)
+		})
 
-			if tagInfo.Length != uint64(len(sendData)) {
-				t.Fatalf("Data length %d != received length %d", len(sendData), tagInfo.Length)
-			}
+		sender.ep, _ = sender.worker.NewEndpoint(epParams)
+		workerAddress.Close()
 
-			request.Close()
-		}})
+		flushRequest, _ := sender.ep.FlushNonBlocking(nil)
 
-	sendRequest, _ := senderEp.SendTagNonBlocking(1, sendMem, uint64(len(sendData)), &UcpRequestParams{
-		Cb: func(request *UcpRequest, status UcsStatus) {
-			if status != UCS_OK {
-				t.Fatalf("Request failed with status: %d", status)
-			}
+		for flushRequest.GetStatus() == UCS_INPROGRESS {
+			sender.worker.Progress()
+			receiver.worker.Progress()
+		}
 
-			request.Close()
-		}})
+		flushRequest.Close()
 
-	for (sendRequest.GetStatus() == UCS_INPROGRESS) || (recvRequest.GetStatus() == UCS_INPROGRESS) {
-		senderWorker.Progress()
-		receiverWorker.Progress()
-	}
+		sendMem := memoryAllocate(sender, uint64(len(sendData)), memType.senderMemType)
+		memorySet(sender, []byte(sendData))
 
-	if recvString := string(GoBytes(receiveMem, uint64(len(sendData)))); recvString != sendData {
-		t.Fatalf("Send data %s != recv data %s", sendData, recvString)
-	}
+		receiveMem := memoryAllocate(receiver, 4096, memType.recvMemType)
```

Comment:
should data size be some kind of parameter?

Reply:
It's a unit test, just to make sure code is working. Later I'll make a benchmark perf test with different configurable parameters.